### PR TITLE
resolved: beef up domain routing

### PR DESCRIPTION
**Pull request by poettering**
Fixes for #10125 #10830 #9825 #9472

### CODE REVIEW

File: man/systemd-resolved.service.xml [hunk 1]
```diff
@@ -132,16 +143,43 @@
...
-    interfaces.</para>
+    <para>Routing of lookups may be influenced by configuring per-interface domain names and other settings. See
+    <citerefentry><refentrytitle>systemd.network</refentrytitle><manvolnum>5</manvolnum></citerefentry> and
+    <citerefentry><refentrytitle>resolvectl</refentrytitle><manvolnum>1</manvolnum></citerefentry> for details. The
+    following query routing logic applies for unicast DNS traffic:</para>
+
+    <itemizedlist>
+      <listitem><para>If a name to look up matches any of the configured search or route-only domains of any link (or
+      the globally configured DNS settings), the "best matching" search/route-only domain is determined: the matching
+      one with the most labels. The query is then sent to all DNS servers of any links or the globally configured DNS
```
Comment (grooverdan) [inline man/systemd-resolved.service.xml]: "If no answer, the query is then sent.."

File: man/systemd-resolved.service.xml [hunk 2]
```diff
@@ -132,16 +143,43 @@
...
+      servers configured with an equally well matching search/route-only domain.</para></listitem>
+
+      <listitem><para>If a query does not match any configured search/route-only domain (neither per-link nor global),
+      it is sent to all DNS servers that are configured on links with the "DNS default route" option set, as well as
+      the globally configured DNS server.</para></listitem>
+
+      <listitem><para>If there is no link configured as "DNS default route" and no global DNS server configured, the
+      compiled-in fallback DNS server is usd.</para></listitem>
```
Comment (grooverdan) [inline man/systemd-resolved.service.xml]: typo: usd

File: man/systemd-resolved.service.xml [hunk 3]
```diff
@@ -132,16 +143,43 @@
...
+
+      <listitem><para>Otherwise the query is failed as no suitable DNS servers could be determined.</para></listitem>
+    </itemizedlist>
+
+    <para>The "DNS default route" option is a boolean setting configureable with <command>resolvectl</command> or in
+    <filename>.network</filename> files. If not set it is implicitly determined based on the configured DNS domains for
```
Comment (grooverdan) [inline man/systemd-resolved.service.xml]: grammar 'If not set, it is...'

File: man/systemd-resolved.service.xml [hunk 4]
```diff
@@ -132,16 +143,43 @@
...
+    a link: if there's any route-only domain (not matching <literal>~.</literal>) it defaults to false, otherwise to
+    true.</para>
+
+    <para>Effectively this means: in order to preferably route all DNS queries not explicitly matched by
+    search/route-only domain configuration to a specific link, configure a <literal>~.</literal> route-only domain on
+    it — this will ensure that other links will not be considered for the queries (unless they too carry such a
```
Comment (grooverdan) [inline man/systemd-resolved.service.xml]: Readability suggestion - break sentence at ' - '

File: src/resolve/resolved-dns-scope.c [hunk 5]
```diff
@@ -441,9 +443,40 @@ int dns_scope_socket_tcp(DnsScope *s, int family, const union in_addr_union *add
...
+
+DnsScopeMatch dns_scope_good_domain(
+                DnsScope *s,
+                int ifindex,
+                uint64_t flags,
+                const char *domain) {
+
         DnsSearchDomain *d;
 
+        /* This returns he the following return values:
```
Comment (grooverdan) [inline src/resolve/resolved-dns-scope.c]: typo 'he'

**Comment by grooverdan** [comment]
tested on fc29 - resolves #10830 #9825 for me.

**Comment by yuwata** [comment]
networkd-test.py fails...

File: man/systemd-resolved.service.xml (hunk 1 above)
Comment (poettering) [reply man/systemd-resolved.service.xml]: oh, this is actually not what is meant here. We first determine the "best matching" domain, and then look for all links that have this "best matching" domain among their configured search domains. The query is then sent to all of them in parallel. This might be a single link or multiple, if the admin specified the same domains on two interfaces.

I will reword this in the man page

**Comment by poettering** [comment]
I force pushed a new version, rebase on current git, with the CI tests (hopefully) fixed. Also fixed the typos and stuff @grooverdan's discovered. PTAL

**Comment by grooverdan** [comment]
I've been running this branch of systemd resolved on my laptop for the last 8 days with the correct domain routing behaviour observed switching between wired and wireless networks, with and without vpn, every few days. Except for occasionally needing to 'resolvectl revert {previous link}' its been working well.

**Comment by evverx** [comment]
Travis CI failed due to https://github.com/systemd/systemd/issues/11170.

File: man/systemd-resolved.service.xml [hunk 6]
```diff
@@ -143,16 +143,45 @@
...
+    <citerefentry><refentrytitle>resolvectl</refentrytitle><manvolnum>1</manvolnum></citerefentry> for details. The
+    following query routing logic applies for unicast DNS traffic:</para>
+
+    <itemizedlist>
+      <listitem><para>If a name to look up matches (that is: is equal to or has as suffix) any of the configured search
+      or route-only domains of any link (or the globally configured DNS settings), the "best matching"
+      search/route-only domain is determined: the matching one with the most labels. The query is then sent to all DNS
+      servers of any links or the globally configured DNS servers associated with this "best matching"
+      search/route-only domain. (Note that more than one link might have this same "best matching" search/route-only
+      domain configured, in which case the query is sent to all of them in parallel).</para></listitem>
```
Comment (keszybz) [inline man/systemd-resolved.service.xml:157]: Yep, this sounds very logical and usable.

File: src/resolve/resolved-dns-scope.c [hunk 7]
```diff
@@ -459,6 +459,21 @@ int dns_scope_socket_tcp(DnsScope *s, int family, const union in_addr_union *add
         return dns_scope_socket(s, SOCK_STREAM, family, address, server, port, ret_socket_address);
 }
 
+static DnsScopeMatch accept_link_local_reverse_lookups(const char *domain) {
+        assert(domain);
+
+        if (dns_name_endswith(domain, "254.169.in-addr.arpa") > 0)
+                return DNS_SCOPE_YES_BASE + 4; /* 4 labels match */
+
+        if (dns_name_endswith(domain, "8.e.f.ip6.arpa") > 0 &&
```
Comment (keszybz) [inline src/resolve/resolved-dns-scope.c]: `||`

File: src/resolve/resolved-dns-scope.c [hunk 8]
```diff
@@ -1393,3 +1391,31 @@ int dns_scope_remove_dnssd_services(DnsScope *scope) {
 
         return 0;
 }
+
+bool dns_scope_has_route_only_domains(DnsScope *scope) {
+
+        DnsSearchDomain *domain, *first;
+        bool route_only = false;
+
+        /* Check if the scope has any route-only domains except of "~.", i. e. whether it should only be used for particular
```
Comment (keszybz) [inline src/resolve/resolved-dns-scope.c]: except for ...

Those comments are also annoying wide.

File: src/resolve/resolved-dns-scope.c [hunk 9]
```diff
@@ -1393,3 +1391,31 @@ int dns_scope_remove_dnssd_services(DnsScope *scope) {
 
         return 0;
 }
+
+bool dns_scope_has_route_only_domains(DnsScope *scope) {
+
```
Comment (keszybz) [inline src/resolve/resolved-dns-scope.c]: Spurious newline.

File: src/resolve/resolved-dns-scope.c [hunk 10]
```diff
@@ -1393,3 +1391,31 @@ int dns_scope_remove_dnssd_services(DnsScope *scope) {
...
+        if (scope->link)
+                first = scope->link->search_domains;
+        else
+                first = scope->manager->search_domains;
+
+        LIST_FOREACH(domains, domain, first) {
+                /* "." means "any domain", thus the interface takes any kind of traffic. */
+                if (dns_name_is_root(DNS_SEARCH_DOMAIN_NAME(domain)))
+                        return false;
```
Comment (keszybz) [inline src/resolve/resolved-dns-scope.c:1417]: Unless I'm misunderstanding the comment above, this does not match the comment. Should this be `continue` instead to match the comment?

The comments in users of this are:
`"Scope of DNS server %s has only route-only domains, not using as global name server"`
`/* Otherwise check if we have any route-only domains: if so, let's not volunteer as default route. */`
They seem to be incompatible :(

File: src/libsystemd/sd-network/sd-network.c [hunk 11]
```diff
@@ -204,6 +204,25 @@ _public_ int sd_network_link_get_route_domains(int ifindex, char ***ret) {
         return network_link_get_strv(ifindex, "ROUTE_DOMAINS", ret);
 }
 
+_public_ int sd_network_link_get_dns_default_route(int ifindex) {
```
Comment (keszybz) [inline src/libsystemd/sd-network/sd-network.c:207]: get → is ? "get" sounds too much like something more than yes/no would be returned.

File: src/resolve/resolved-link.c [hunk 12]
```diff
@@ -301,6 +301,27 @@ static int link_update_dns_servers(Link *l) {
...
+        }
+        if (r < 0)
+                goto clear;
+
+        l->default_route = r > 0;
+        return 0;
+
+clear:
+        l->default_route = -1;
+        return r;
```
Comment (keszybz) [inline src/resolve/resolved-link.c:322]: Hmm, resetting the value even in the error path is unexpected.

File: src/resolve/resolved-link.c [hunk 13]
```diff
@@ -617,6 +638,10 @@ static void link_read_settings(Link *l) {
         r = link_update_search_domains(l);
         if (r < 0)
                 log_warning_errno(r, "Failed to read search domains for interface %s, ignoring: %m", l->name);
+
+        r = link_update_default_route(l);
+        if (r < 0)
+                log_warning_errno(r, "failed to read default route setting for interface %s, ignoring: %m", l->name);
```
Comment (keszybz) [inline src/resolve/resolved-link.c]: "Failed"

Also, "ignoring" is not entirely appropriate, since it's being reset anyway.

File: man/systemd.network.xml [hunk 14]
```diff
@@ -547,6 +547,17 @@
             name servers limited to a specific link.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>DNSDefaultRoute=</varname></term>
+          <listitem>
+            <para>Take a boolean argument. If true, this link's configured DNS servers are used for resolving domain
```
Comment (keszybz) [inline man/systemd.network.xml]: "Takes"

File: man/systemd.network.xml [hunk 15]
```diff
@@ -547,6 +547,17 @@
...
+            names that do not match any link's configured <varname>Domains=</varname> setting. If false, this link's
+            configured DNS servers are never used for such domains, and are exclusively used for resolving names that
+            match at least one of the domains configured on this link. If not specified defaults to an automatic mode:
+            if any routing domains are configured on the link (see above) queries not matching any link's configured
```
Comment (keszybz) [inline man/systemd.network.xml]: Please insert a comma after "(see above)".

File: man/systemd.network.xml [hunk 16]
```diff
@@ -547,6 +547,17 @@
...
+            domains will not be routed to this link, otherwise they will.</para>
```
Comment (keszybz) [inline man/systemd.network.xml]: Maybe reword this as "If not specified defaults to an automatic mode: queries no matching any link's configured domains will be routed to this link if it has no routing domains configured."

File: src/resolve/resolved-dns-scope.c (hunk 7 above)
Comment (poettering) [reply src/resolve/resolved-dns-scope.c]: OUCH!

File: src/resolve/resolved-link.c (hunk 12 above)
Comment (poettering) [reply src/resolve/resolved-link.c:322]: it's how all the other link_update_xyz() calls operate too: if we can't read the data from networkd we'll unset the options... i.e. parse errors invalidate data, they don't mean old data remains. And I'd argue it's nice to do it that way since it means we expose the same behaviour if resolved is just started up and reads networkd data for the first time, and if it is already running and the data is corrupted while it is operating, since in both cases it will now invalidate the data. If we'd leave the old data in place then the behaviour would differ between those two cases, if you follow what i mean...

File: src/resolve/resolved-link.c (hunk 13 above)
Comment (poettering) [reply src/resolve/resolved-link.c]: will change to "proceeding anyway"

File: src/resolve/resolved-dns-scope.c (hunk 8 above)
Comment (poettering) [reply src/resolve/resolved-dns-scope.c]: (this comment is actually removed later on again in the same PR...)

File: man/systemd.network.xml (hunk 15 above)
Comment (poettering) [reply man/systemd.network.xml]: (i used your suggested rephrasing, below, which drops the see above anyway, so i guess this doesn't apply anymore)

File: src/resolve/resolved-dns-scope.c (hunk 10 above)
Comment (poettering) [reply src/resolve/resolved-dns-scope.c:1417]: So, this is not actually new code, but just some minor reordering of the old one, behaviour is the same.

I think the automatic logic of "let's not use this link as default route if it has any route-only domains set" was not necessarily the greatest logic ever, but it's also not totally flawed, and maybe was OK without this new explicit bool flag this patch set adds. So my intention was to retain the original logic as implicit default, already for compat reasons, and it works the way it works after all..

So this early exit and returning "false" even makes sense: after all the call is supposed to discern two kinds of links:

1. links that shall get only lookups for specific domains (i.e. your vpn link)
2. links that shall get any lookups (i.e. your internet link)

In the first case we return true, in the second case false.

Now, if you have "~." on as search domain on a link, then this should rightfully be considered of type 2. Hence the check does the right thing. After all if you have "~." on a link, then it doesn't really matter what other search domains it has too much, after all you explicitly said: "send me everything" that way...

Hope this makes sense?

I will extend the comment a bit more, about this.

File: src/libsystemd/sd-network/sd-network.c (hunk 11 above)
Comment (poettering) [reply src/libsystemd/sd-network/sd-network.c:207]: So I thought about this, but we have sd_network_link_get_required_for_online() already which also returns a boolean, so I stuck to the same naming logic there.

THis stuff isn't exported, I figured, and it wasn't clear to me that "is" is really that better here, as after all the link "is" not necessarily the DNS default route, but it is just supposed to be, from networkd's pov...

Anyway, so I was in doubt, and it's private, hence I figured we should stick to the current naming scheme. If we make sd-network public we can of course revisit this...

**Comment by poettering** [comment]
I force pushed a new version now. Only changes are the ones @keszybz pointed out above, modulo the ones I commented out above. I also rebased, in the hope the CI goes all green. PTAL!

File: src/libsystemd/sd-network/sd-network.c (hunk 11 above)
Comment (keszybz) [reply src/libsystemd/sd-network/sd-network.c:207]: Ack.

File: src/resolve/resolved-link.c (hunk 12 above)
Comment (keszybz) [reply src/resolve/resolved-link.c:322]: OK, makes sense.

**Comment by poettering** [comment]
OK, let's merge this. The changes I made were mostly in docs after all, and @keszybz liked it pretty much, the CI is all green, @grooverdan tested it successfully.

**Comment by nathan-wmrln** [comment]
There's a special case that's been biting me: I have an interface that looks like this:
```
Link 8 (tun0)
      Current Scopes: DNS
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
         DNS Servers: 172.31.0.2
          DNS Domain: ~.
```
Everything works fine when /etc/resolv.conf is symlinked to /run/systemd/resolve/stub-resolv.conf. But in /run/systemd/resolve/resolv.conf, the nameservers that I don't want to use, still appear, in an order that seems nondeterministic:

```
nameserver x.y.z.a
nameserver y.z.a.b
nameserver 172.31.0.2
```
It would be best if it could either ensure that 172.31.0.2 is always listed first in such a special case, or, better yet to avoid DNS leaks over a VPN, not at all. Is that case dealt with by this PR?

**Comment by tomtastic** [comment]
I have the same issue as @nathan-wmrln using systemd 240. With `/etc/systemd/network/all.network` :
```
[DHCP]
UseDNS
```
works for `/run/systemd/resolve/stub-resolv.conf`, but not for `/run/resolvconf/resolv.conf`
